nsxt/resource_nsxt_policy_ipsec_vpn_session.go

```diff
@@ -0,0 +1,623 @@
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/bindings"
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/data"
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/protocol/client"
+
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt/infra/tier_0s/locale_services/ipsec_vpn_services"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt/model"
+)
+
+var IPSecVpnSessionResourceType = []string{
+	model.IPSecVpnSession_RESOURCE_TYPE_POLICYBASEDIPSECVPNSESSION,
+	model.IPSecVpnSession_RESOURCE_TYPE_ROUTEBASEDIPSECVPNSESSION,
+}
+
+var IPSecVpnSessionAuthenticationMode = []string{
+	model.IPSecVpnSession_AUTHENTICATION_MODE_PSK,
+	model.IPSecVpnSession_AUTHENTICATION_MODE_CERTIFICATE,
+}
+
+var IPSecVpnSessionConnectionInitiationMode = []string{
+	model.IPSecVpnSession_CONNECTION_INITIATION_MODE_INITIATOR,
+	model.IPSecVpnSession_CONNECTION_INITIATION_MODE_RESPOND_ONLY,
+	model.IPSecVpnSession_CONNECTION_INITIATION_MODE_ON_DEMAND,
+}
+var IPSecVpnSessionComplianceSuite = []string{
+	model.IPSecVpnSession_COMPLIANCE_SUITE_CNSA,
+	model.IPSecVpnSession_COMPLIANCE_SUITE_SUITE_B_GCM_128,
+	model.IPSecVpnSession_COMPLIANCE_SUITE_SUITE_B_GCM_256,
+	model.IPSecVpnSession_COMPLIANCE_SUITE_PRIME,
+	model.IPSecVpnSession_COMPLIANCE_SUITE_FOUNDATION,
+	model.IPSecVpnSession_COMPLIANCE_SUITE_FIPS,
+	model.IPSecVpnSession_COMPLIANCE_SUITE_NONE,
+}
+
+var IPSecRulesActionValues = []string{
+	model.IPSecVpnRule_ACTION_PROTECT,
+	model.IPSecVpnRule_ACTION_BYPASS,
+}
+
+func resourceNsxtPolicyIPSecVpnSession() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceNsxtPolicyIPSecVpnSessionCreate,
+		Read:   resourceNsxtPolicyIPSecVpnSessionRead,
+		Update: resourceNsxtPolicyIPSecVpnSessionUpdate,
+		Delete: resourceNsxtPolicyIPSecVpnSessionDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"nsx_id":              getNsxIDSchema(),
+			"path":                getPathSchema(),
+			"display_name":        getDisplayNameSchema(),
+			"description":         getDescriptionSchema(),
+			"revision":            getRevisionSchema(),
+			"tag":                 getTagsSchema(),
+			"tunnel_profile_path": getPolicyPathSchema(true, false, "Policy path referencing Tunnel profile to be used."),
+			"local_endpoint_path": getPolicyPathSchema(true, false, "Policy path referencing Local endpoint."),
+			"ike_profile_path": {
+				Type:        schema.TypeString,
+				Description: "Policy path referencing IKE profile.",
+				Optional:    true,
+			},
+			"dpd_profile_path": {
+				Type:        schema.TypeString,
+				Description: "Policy path referencing Dead Peer Detection (DPD) profile. Default is set to system default profile.",
+				Optional:    true,
+				Computed:    true,
+			},
+			"vpn_type": {
```

Comment:
We try to use human-friendly strings in the provider, and `RouteBasedIPSecVpnSession` is too long, people will curse us :) I'd suggest to go with `RouteBased` and `PolicyBased`, even if it means a bit more code.